### PR TITLE
Initial Changes For video aspect ratio changes

### PR DESCRIPTION
**Pull request by Akshay-1024**
https://github.com/recloudstream/cloudstream/issues/379

Made changes related to aspect ratio in the AbstractPlayerFragment.kt to handle for standard aspect ratios

Link for the video of Pop Up
https://drive.google.com/file/d/1frhCkOWAaTo3n1yJEPEYhnsaAIlfYhDy/view?usp=sharing

### Updated ScreenShot using internal SingleSelectionHelper

image

### CODE REVIEW

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (CranberrySoup) [inline app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:645]: Avoid unchecked casts in Android, they will crash if not satisfied. Use `else if (type is Int)` for the exact same effect, but more future proof

File: app/src/main/res/values/strings.xml
Comment (CranberrySoup) [inline app/src/main/res/values/strings.xml:436]: I dont know if these are placeholders, but if they are not, then "Widescreen 16:9" and "Square 1:1" might be more readable and consistent

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (CranberrySoup) [inline app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:639]: With this many options a popup might be preferable over pressing resize 9 times, but this is a great first framework!

**Review by CranberrySoup — COMMENTED** [review]
Good job!
It's an early review so you might already be working on the details I mentioned, then you can just dismiss those comments.

You solved the aspect ratio problem very nicely :+1:

**Comment by C10udburst** [comment]
btw, if you want to indicate you are still working on a PR and don't want it merged yet its good practise to mark it as draft. there should be a button on the right hand side to do so

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (Akshay-1024) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:645]: Yea, this is correct. I have Updated this code in my local

File: app/src/main/res/values/strings.xml
Comment (Akshay-1024) [reply app/src/main/res/values/strings.xml:436]: Updated them as needed

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (Akshay-1024) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt]: I have made changes for the pop up, Attaching the video in the PR for reference let me know if I have to make any changes or suggestions

**Comment by Akshay-1024** [comment]
> btw, if you want to indicate you are still working on a PR and don't want it merged yet its good practise to mark it as draft. there should be a button on the right hand side to do so

Yes, I would follow that from the next time. Thanks.

**Review by CranberrySoup — COMMENTED** [review]
Looks good code wise, however I have not had time to test it yet

**Comment by fire-light42** [comment]
Use SingleSelectionHelper.showDialog to show a dialog in the cs3 theme.

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (Akshay-1024) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt]: Using SingleSelectionHelper.showDialog i am failing it silently.

**Comment by Akshay-1024** [comment]
> Use SingleSelectionHelper.showDialog to show a dialog in the cs3 theme.

Thanks Used it.

**Comment by Akshay-1024** [comment]
Hi @C10udburst , @CranberrySoup , @fire-light42 any update on this?????

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (CranberrySoup) [inline app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:630]:
```suggestion
        val aspectRatioOptions = PlayerResize.entries.map { getString(it.nameRes) }
```
Better to use the recommended replacement for values

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (CranberrySoup) [inline app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:641]:
```suggestion
                val selectedOption = PlayerResize.entries[selectedIndex]
```

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (CranberrySoup) [inline app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:634]:
```suggestion
            DataStoreHelper.resizeMode,
```
Make the selected index what the user has previously picked

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (CranberrySoup) [inline app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:669]:
```suggestion
            playerView?.resizeMode = type
            aspectRatioFrameLayout?.setAspectRatio(0f) // Reset the aspect ratio when using built in resizing
```
Without this the implementation is bugged. This can easily be checked by switching from "Square" to "Fit to screen"

**Review by CranberrySoup — CHANGES_REQUESTED** [review]
The bug is critical and the index should be fixed for app-wide consistency. You can just click and accept the changes.

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (Akshay-1024) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:669]: Yes, I am also getting this while switching from Square to fit to screen, but after adding the setAspectRatio(0f) , after testing it was missing when switching from stretch to zoom, and other custom aspect ratio to predefined like fit, fill, zoom all behaving like fill. I am looking into that.

Comment (Akshay-1024) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:669]: Once I will try this on real device

Comment (Akshay-1024) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:669]: There maybe some issue in my code I will resolve it. I do agree with the idea setAspectRatio(0f) should fix the issue.

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (C10udburst) [inline app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:673]: Since after this change the aspect ratio is explicitly set by the user, it may be worth thinking if this toast is necessary.

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (fire-light42) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:669]: @Akshay-1024 I cannot merge without a fix to this bug. Should I apply the fix or are you working on something?

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (Akshay-1024) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt]: Removed Toast from the Resize function.

File: app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt
Comment (Akshay-1024) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:664]: @fire-light42 I have removed fit as its working same as widescreen, I have removed Zoom as well because after the changes Zoom is basically behaving same as fill So Kept only Fill for predefined and rest all are custom aspect ratios rest all are working fine, but if you have any fix for zoom as well let me know. I have Pushed the changes

Comment (fire-light42) [reply app/src/main/java/com/lagradost/cloudstream3/ui/player/AbstractPlayerFragment.kt:664]: I cant merge this if you have removed Fit and Zoom, all three original options act differently depending on screen size and video size. Just because they look the same on your phone, does not mean they function identically.

**Review by fire-light42 — CHANGES_REQUESTED** [review]
After testing the code, I can say that this code is deeply flawed. The concept of overriding the aspect ratio via setAspectRatio overrides the aspect ratio information set by the playerView means that we cant recover the aspect ratio for RESIZE_MODE_FILL RESIZE_MODE_FIT and RESIZE_MODE_ZOOM if we choose to go back. 

By setting aspect ratio to 0.0f we dont fix anything as the aspect ratio view wont render using the resizemode. 

AspectRatioFrameLayout::onMeasure
```kotlin
if (videoAspectRatio <= 0) {
      // Aspect ratio not set.
      return;
}
``` 

Futhermore, the code is competing with contentFrame.setAspectRatio from playerview so even if you store the variables locally, the player might decide to override that aspect ratio as well.

**Pull request closed without merging**
